### PR TITLE
Fixed #31623 -- Allowed specifying number of adjacent time units in timesince()/timeuntil().

### PR DESCRIPTION
**Pull request by timpark0807**
### **Overview**
**Ticket:** https://code.djangoproject.com/ticket/31623

The `timesince` function returns the time between two datetime objects as a formatted string.

The previous implementation always returned two units of time. This update improves flexibility by allowing the user to specify the number of units to be returned. To preserve existing functionality, the default depth value is set to 2. 

### **Example**
 - depth = 1 returns "2 weeks"
 - depth = 2 returns "2 weeks, 3 days"
 - depth = 3 returns "2 weeks, 3 days, 3 hours"
 - depth = 4 returns "2 weeks, 3 days, 3 hours, 13 minutes"

### **Update includes:**
- New depth functionality 
- New test cases 
- Updated Doc Strings
- Updated Release Notes 

Let me know if you see anything.

Thanks!

### CODE REVIEW

**Comment by timpark0807** [comment]
PR ready to review.

File: django/utils/timesince.py [hunk 1]
```diff
@@ -54,6 +59,44 @@ def timesince(d, now=None, reversed=False, time_strings=None):
...
+    Returns the number of seconds elapsed between two datetime objects.
+    """
+    delta = calculate_delta(now, d)
+    return delta.days * 24 * 60 * 60 + delta.seconds
+
+
+def calculate_delta(now, d):
+    """
+    Returns a datetime.timedelta object representing elapsed time between two datetime objects.
+    """
```
Comment (claudep) [inline django/utils/timesince.py:99]: What's the interest of creating separate `calculate` functions while they seem to be called only once?

Comment (timpark0807) [reply django/utils/timesince.py:99]: Hey Claude! 

The functions were meant to make the code more readable and self-documenting. There was a discussion on the forum post linked below about extracting the logic for calculating the time before formatting. This idea was mentioned in the ticket as well. 

Let me know your thoughts.  

Discussion link: https://groups.google.com/forum/#!topic/django-developers/pEaX5X-wpWo

Comment (claudep) [reply django/utils/timesince.py:99]: I see the idea, but for me if a function is called only once and only contains some simple lines, the function call overhead is not worth it. You can let this for now and wait for the Django fellows opinion.

Comment (timpark0807) [reply django/utils/timesince.py:99]: Sounds good. I'm open to refactoring as necessary. Thanks for the input!

**Comment by smithdc1** [comment]
Hi @timpark0807 

Thanks for this.  👍 

My notes:

- What do you think about `timeuntil`. Could pass through depth here as well?

- https://docs.djangoproject.com/en/3.0/ref/templates/builtins/#timesince These docs will need updating.

- This doesn't quite work as I would have expected in the example below. In the second example I'd have expected a number of days to be returned. Thoughts? 

``` python
>>> d = datetime.datetime(year=2018, month=3, day=22)
>>> timesince(d, depth=4)
'2\xa0years, 3\xa0months, 2\xa0weeks, 2\xa0days'

>>> d = datetime.datetime(year=2018, month=3, day=2)
>>> timesince(d, depth=4)
'2\xa0years, 4\xa0months'
```

**Comment by timpark0807** [comment]
> Hi @timpark0807
> 
> Thanks for this. 👍
> 
> My notes:
> 
> * What do you think about `timeuntil`. Could pass through depth here as well?
> * https://docs.djangoproject.com/en/3.0/ref/templates/builtins/#timesince These docs will need updating.
> * This doesn't quite work as I would have expected in the example below. In the second example I'd have expected a number of days to be returned. Thoughts?
> 
> ```python
> >>> d = datetime.datetime(year=2018, month=3, day=22)
> >>> timesince(d, depth=4)
> '2\xa0years, 3\xa0months, 2\xa0weeks, 2\xa0days'
> 
> >>> d = datetime.datetime(year=2018, month=3, day=2)
> >>> timesince(d, depth=4)
> '2\xa0years, 4\xa0months'
> ```

@smithdc1 Thanks for taking the time to look through this PR. 

My responses are below: 

- Great catch on the `timeuntil` depth. I'll modify that to accept a pass through depth parameter.   
- Sounds good. I will work on putting those docs together once this gets approval from the community to go through. 
- The timesince function only returns **adjacent** time units. Because there are 0 weeks between 3/2/18 and 7/5/20, the day/hour/minute units will not be processed.  

Screen Shot 2020-07-05 at 9 52 39 AM

Screenshot using http://www.mathcats.com/explore/elapsedtime.html

As a side note, I was able to implement a timesince function that would be able to skip through missing adjacent units (The example above would return `2 Years, 4 Months, 3 Days, 10 Hours`) but I didn't want to mess with original functionality. I'm happy to discuss this further and create a new ticket if this type of behavior is desired.

**Comment by smithdc1** [comment]
> * The timesince function only returns **adjacent** time units. Because there are 0 weeks between 3/2/18 and 7/5/20, the day/hour/minute units will not be processed.

The current function only returns a depth of two, so stopping if there isn't an adjacent time depth gives the answer I would expect. (example below)

When depth is increased to three or above there is a chance one of the depths is "0" but there are values at more granular time depths. Chopping at the first "0" I find surprising (just my opinion, I'm sure there are others!). 

For completeness I think the example you posted is what would expect. 👍 
This --> `2 Years, 4 Months, 3 Days, 10 Hours`
Not --> `2 Years, 4 Months, 0 Weeks, 3 Days, 10 Hours`

``` python
>>> d = datetime(year=2018, month=6, day=2)
>>> timesince(d)
'2\xa0years, 1\xa0month'
>>> d = datetime(year=2018, month=6, day=20)
>>> timesince(d)
'2\xa0years'
```

**Comment by timpark0807** [comment]
> When depth is increased to three or above there is a chance one of the depths is "0" but there are values at more granular time depths. Chopping at the first "0" I find surprising (just my opinion, I'm sure there are others!).
> 
> For completeness I think the example you posted is what would expect. 👍
> This --> `2 Years, 4 Months, 3 Days, 10 Hours`
> Not --> `2 Years, 4 Months, 0 Weeks, 3 Days, 10 Hours`

I 100% agree with your opinion. 

However, I decided against changing the behavior within this PR because the original ticket specifically states `"The existing rule of values having to be adjacent to one another should still remain."` 

Let's get the community opinion to see if we should proceed with a modified ticket to **not** chop at the first "0".

File: django/utils/timesince.py (hunk 1 above)
Comment (felixxm) [reply django/utils/timesince.py:99]: I agree with Claude.

Comment (timpark0807) [reply django/utils/timesince.py]: @felixxm @claudep 

Hey guys - Per feedback, I removed the separate `calculate` functions in the latest commit 👍

File: docs/releases/3.2.txt [hunk 2]
```diff
@@ -357,6 +357,10 @@ Miscellaneous
   :attr:`~django.db.models.Model._default_manager` to check that related
   instances exist.
 
+* The ``~django.utils.timesince.timesince`` function now allows the use of
```
Comment (felixxm) [inline docs/releases/3.2.txt:360]:
```
* The ``django.utils.timesince.timesince()`` and ``django.utils.timesince.timeuntil()`` functions now allow ....
```

File: tests/utils_tests/test_timesince.py [hunk 3]
```diff
@@ -140,3 +140,42 @@ def test_thousand_years_ago(self):
         t = datetime.datetime(1007, 8, 14, 13, 46, 0)
         self.assertEqual(timesince(t, self.t), '1000\xa0years')
         self.assertEqual(timeuntil(self.t, t), '1000\xa0years')
+
+    def test_depth(self):
+        t = self.t + self.oneyear + self.onemonth + self.oneweek + self.oneday + self.onehour
+        self.assertEqual(timesince(self.t, t, depth=5), '1\xa0year, 1\xa0month, 1\xa0week, 1\xa0day, 1\xa0hour')
```
Comment (felixxm) [inline tests/utils_tests/test_timesince.py:146]: You can merge `test_depth()`, `test_depth_out_of_range()`, and `test_depth_one()` into a single test `test_timesince_depth()` and use `self.subTest()`.

File: django/utils/timesince.py [hunk 4]
```diff
@@ -24,20 +24,25 @@
...
 
-def timesince(d, now=None, reversed=False, time_strings=None):
+def timesince(d, now=None, reversed=False, time_strings=None, depth=2):
     """
-    Take two datetime objects and return the time between d and now as a nicely
-    formatted string, e.g. "10 minutes". If d occurs after now, return
-    "0 minutes".
+    Take two datetime objects and return the time between d and now as a formatted string.
 
+    If d occurs after now, return "0 minutes".
```
Comment (felixxm) [inline django/utils/timesince.py:31]: We can revert this unrelated change.

File: django/utils/timesince.py [hunk 5]
```diff
@@ -24,20 +24,25 @@
...
-    Seconds and microseconds are ignored.  Up to two adjacent units will be
-    displayed.  For example, "2 weeks, 3 days" and "1 year, 3 months" are
-    possible outputs, but "2 weeks, 3 hours" and "1 year, 5 days" are not.
+    Seconds and microseconds are ignored.
 
-    `time_strings` is an optional dict of strings to replace the default
-    TIME_STRINGS dict.
+    Only adjacent units will be displayed.
+        Valid   : "2 weeks, 3 days" and "1 year, 3 months"
+        Invalid : "2 weeks, 3 hours" and "1 year, 5 days"
```
Comment (felixxm) [inline django/utils/timesince.py:37]: I would revert this change.

File: django/utils/timesince.py [hunk 6]
```diff
@@ -66,26 +71,34 @@ def timesince(d, now=None, reversed=False, time_strings=None):
     delta -= datetime.timedelta(leapdays)
 
     # ignore microseconds
-    since = delta.days * 24 * 60 * 60 + delta.seconds
-    if since <= 0:
+    total_seconds_since = delta.days * 24 * 60 * 60 + delta.seconds
+    if total_seconds_since <= 0:
```
Comment (felixxm) [inline django/utils/timesince.py:75]: Changing the names of variables which is not necessary for this patch makes it harder to review, can we revert them? e.g. `total_seconds_since -> since`, or 
```
for index, (seconds_per_chunk, chunk_name) in enumerate(TIMESINCE_CHUNKS):
```
->
```
for i, (seconds, name) in enumerate(TIMESINCE_CHUNKS):
```

File: django/utils/timesince.py [hunk 7]
```diff
@@ -24,20 +24,25 @@
...
+        Valid   : "2 weeks, 3 days" and "1 year, 3 months"
+        Invalid : "2 weeks, 3 hours" and "1 year, 5 days"
 
+    `depth` is an optimal integer to change the number of adjacent units to display.
+        depth = 1 returns "2 weeks"
+        depth = 2 returns "2 weeks, 3 days"
+        depth = 3 returns "2 weeks, 3 days, 3 hours"
+        depth = 4 returns "2 weeks, 3 days, 3 hours, 13 minutes"
+
+    `time_strings` is an optional dict of strings to replace the default TIME_STRINGS dict.
```
Comment (felixxm) [inline django/utils/timesince.py:45]: Please revert change in wrapping, and wrap new docstring and comments at 79 chars.

**Comment by timpark0807** [comment]
@felixxm Thanks for the thorough review. I got a bit carried away with my first PR. I'll keep these comments in mind for future commits. Let me know if you see anything else 👍

**Comment by felixxm** [comment]
@claudep You mentioned in the ticket and on the mailing list that the most important thing is to make it easily customizable, however I don't see how this PR solves this. I'm fine with limiting this PR to the `depth` argument. Do you think we should refactor `timesince()`?

**Comment by claudep** [comment]
I don't remember if I had a specific idea in mind sorry. I guess a refactor can always be done later.

**Comment by felixxm** [comment]
@timpark0807 Thanks :+1: I added check for a `depth` value and pushed small edits. IMO we should keep the current behavior and refactoring (if necessary) can be done in a separate PR/ticket.

**Comment by Edumap** [comment]
can you help me tp develop [how old am i](http://howoldami.today/) script similar to this?
